core/src/main/java/ai/timefold/solver/core/impl/bavet/NodeNetwork.java

```diff
@@ -29,23 +30,27 @@ public int layerCount() {
         return layeredNodes.length;
     }
 
-    @SuppressWarnings("unchecked")
-    public AbstractForEachUniNode<Object>[] getApplicableForEachNodes(Class<?> factClass) {
+    public Stream<AbstractForEachUniNode<?>> getForEachNodes() {
+        return declaredClassToNodeMap.values()
+                .stream()
+                .flatMap(List::stream);
+    }
+
+    public Stream<AbstractForEachUniNode<?>> getForEachNodes(Class<?> factClass) {
         return declaredClassToNodeMap.entrySet()
                 .stream()
                 .filter(entry -> entry.getKey().isAssignableFrom(factClass))
                 .map(Map.Entry::getValue)
-                .flatMap(List::stream)
-                .toArray(AbstractForEachUniNode[]::new);
+                .flatMap(List::stream);
     }
 
-    public void propagate() {
+    public void settle() {
```

Comment:
Propagate seems to be a common term used in CS. I'm unsure if the method name change is good idea.

Reply:
The term was actually conflating two different things:
- Propagation, which is what happens to tuples in nodes.
- Settling, which is a process in which a series of nodes have their propagations triggered.

The current situation is clearer now. Once all nodes have propagated, the node network is considered settled.